### DISCUSSION TITLE
What is the purpose of mapArray and when/how to use it?

### DISCUSSION THREAD

**Discussion by zhaoyao91** [solidjs/solid#428] [Q&A] [answered]
I found the `mapArray` method here and there which looks so complicated. I am not sure what's the purpose and best practice of it.

**Comment by ryansolid** [comment]
It's the basis of `<For>`. Sometimes you want to map something from one reactive structure to another and not recreate everything when it updates. It basically allows mapping that will handle not recreating rows, and handle row swaps. In the end, only new rows are created when new data is present.

**Comment by Freak613** [comment, marked as answer]
Alternative definition could be that it is tool to effectively maintain collection of side-effects (if we consider DOM element as side effect). On data arrival, it looks for what should be created and what should be destroyed. It is similar to dom nodes reconcillation (and has similar code inside) with difference that moving elements doesn't have any effect.
It's not limited to DOM nodes, this way could be managed anything that is costly to create or should not be destroyed until item is gone from the source. It can manage object instances, subscriptions, event listeners, sockets, maybe file descriptors, basically everything effectful (I don't know if solid supports it, but initial design can support all of it).

Originally, it was utility to reuse results of previous computation for each item, when array changed. But it maybe not so visible with plain data as it is with some heavyweight stuff.

When to use it: when you have expensive data computation per each item or you want to avoid recreating some heavy entities for each item on each array update.

**Reply by zhaoyao91** [reply]
Try to clarify, if A is the original array, B is the mapped array:

- If there is a new item a added in A, there will be a new item b added in B.
- If item a in A is removed, there is an item b removed from B.
- If the map function invocation for item a used some singal, and the signal invalidates, then the function will rerun, new item b' will be generated to replace the item b in B.
- If item a in A changed position, the corresponding item b in B will change position correspondently.
- All the above happens with modification of B, instead of creating a new B, or touching the items in B which are not needed to update.

Is this understanding right? Or somewhere should be amended?

**Reply by ryansolid** [reply]
Mostly. except point 3:
> * if the map function invocation for item a used some singal, and the signal invalidates, then the function will rerun, new item b' will be generated to replace the item b in B.

mapArray callback map functions are untracked. You can create reactive stuff inside that will re-run but the map function never re-runs. It is created once and then moved or eventually deleted. Reading signals in the map function don't do anything unless wrapped in their own computations. Those nested computations can execute independently without triggering the array reconciliation logic. In so we only need to even run the diff if the array is modified, but not the content in it. This isolation lends to a large part of Solid's scaling performance.

**Reply by zhaoyao91** [reply]
Good explain. Now I understand it. But, under this setting, if we want the output item reflects the input item (as the input item is reactive state object or mutable object), we have to make such code:

```ts

(input) => {
  return {
    get name () {
      return input.name;
    }
  }
}

```

right? Seems not straghtforword.

**Reply by ryansolid** [reply]
Or:
```
(input) => () => input.name

(input) => <>{input.name}</>
```
Depending on your scenario you might want to key by index anyway when dealing with simple primitive values, `indexArray`. Usually you will be using this with nested state created with `createState` or your own objects with their own signals. The problem with using primitive values is changing them essentially is rekeying.. there is no benefit to being keyed.

In the example you gave what do you want the map to return? Like what you wrote makes sense if you want to end up with an array of objects that have a name property or a array of signals etc, but in those cases that would be expected that you mapped to an object or a function. The only difference of if the mapFn was reactive is that you could map to resolved values. Sort of a flat map. Not my usual my use case but I'm sure the helper could be written to function that way. But at that point there wouldn't be granular reactivity so you might as well just write to a single signal array.

Maybe help me understand what you are looking for.

**Reply by zhaoyao91** [reply]
Thanks, I currently didn't have a concrete case, just study the essense :)

**Reply by this-spring** [reply]
It sounds like a concept in some programming languages. Can you tell me specifically whether there is any documentation explaining this concept?
